### ISSUE TITLE
[Enhancement]: 

### ISSUE THREAD

**Issue by inventor2525** [Pythagora-io/gpt-pilot#763]
### Version

Command-line (Python) version

### Suggestion

Please use file diffs for edits.

These long form re-writes of entire files are very costly.

prompts like this are really bad for automated systems:

### implement_changes.prompt
```txt
I need to modify file `{{ file_name }}` that currently looks like this:
\ ```
{{ file_content }}
\ ```

I want you to implement changes described in previous message and show me the COMPLETE NEW VERSION of this file in this format:
-----------------------format----------------------------
\ ```
the full contents of the updated file, without skipping over any content
\ ```
------------------------end_of_format---------------------------

**IMPORTANT**:If the instructions have comments like `// ..add code here...` or `# placeholder for code`, instead of copying the comment, interpret the instructions and output the relevant code.

**IMPORTANT**: Your reply MUST NOT omit any code in the new implementation or substitute anything with comments like `// .. rest of the code goes here ..` or `# insert existing code here`, because I will overwrite the existing file with the content you provide. Output ONLY the content for this file, without additional explanation, suggestions or notes. Your output MUST start with ``` and MUST end with ``` and include only the complete file contents.

**IMPORTANT**: For hardcoded configuration values that the user needs to change, mark the line that needs user configuration with `INPUT_REQUIRED {config_description}` comment,  where `config_description` is a description of the value that needs to be set by the user. Use appropriate syntax for comments in the file you're saving (for example `// INPUT_REQUIRED {config_description}` in JavaScript). NEVER ask the user to write code or provide implementation, even if the instructions suggest it! If the file type doesn't support comments (eg JSON), don't add any.

{{ logs_and_error_handling }}

```

**Comment by senko** [comment]
I agree it's not optimal to have to rewrite whole files.

Turns out LLMs are very bad at producing consistent diffs. We actually [used this approach a few months ago](https://github.com/Pythagora-io/gpt-pilot/blob/cb95ec09bfdb540072683b4082e690d87988ba70/pilot/prompts/development/implement_changes.prompt) and it was so error prone we decided to ditch it and use full rewrites.

Eventually we'll probably need to go with a specialized fine-tuned model just for this (or some popular LLM will get first-class support for diffs).

**Comment by inventor2525** [comment]
Yeah :/ ... I agree. I've looked at using LLMs to produce diffs directly in the past.

I still want to look at it more myself, but I'm currently trying to look at how they could handle terminal emulation for code insertion because of that, but that's a much larger rabbit hole.

I didn't do overly extensive testing, but I do wonder how much better it would do if you instructed it to write patch files instead. git's format is already well explored by llm's, it's a bit less for it to chew on instruction wise. Large context's just generally don't do well. I suspect it would muddy the format at times without guard rails (can only be done with local llm's but basically you cull the log_probs for tokens that meet a format, statically analyze the diff as it's produced with reference to the code it's diffing to guide it, obviously a lot and not ideal but works well), but even Chat-GPT months after it came out surprised me with this. Even when it gets format right though I've watched it get confused on exactly which line needs to be changed (even guard rails don't help that).

It just seems painfully a must to explore further and didn't know how much you did. I know [mentat](https://github.com/AbanteAI/mentat) was trying it (fyi), haven't been keeping up, don't know if they ditched it, but don't know if you 2 are talking.

I did seem to notice that it responded better to fake conversation data than listening to instructions in a single prompt, since that follows it's already more used (dare I say over-fitted) formatting, but that's sent me down a much longer rabbit hole that... Don't know if you tried that but, figure I'd throw it out there.

On training diff specialized LLMs (much agreed!!!), this is a bit out of scope of my original suggestion, and it looks like your db could already help with this (side bar: wish you were using subprocess calls to git than sqlite but, it'd work)... 

   I would hope you are planning to opt in (or very apparently opt out) crowd source LLM code change examples.

If user can say even that a change is good (or you figure it out by processing the conversation with the user), you could generate a LOT of diff examples (or things you can get diffs from given the typical dreaded "// ...rest of code here..." comments ). I mean, we can all git diffs from pulling random high reviewed repos, but it would be good if coding expert LLMs (ones that have simply looked at lots of code like GPT4 or code llama) didn't have to loose their coding capabilities to being over-fit to a diff format. If you can process the output of a specialist that outputs things with "rest of code here" (cheaper and faster!!), it'd obviously be a lot better, but that means having lots of code changes with those comments as examples.

(And yea... I know licensing might highly effect that, but I both wonder if the lawyers wouldn't say: _"They didn't use your LLM's output, they used a HIGHlY post processed version of it to produce diff examples, and you said the content generated by the LLM was "theirs". You learn with your LLM on other's LLM output (since it all goes online now corrupting everyone's data) anyway."_..... but.... If you don't want to go down that rabbit hole? at least scrape the open model data or track which model produced any given data and follow that models license on a case by case)

Kinda wish there was a [Open Assistant](https://github.com/LAION-AI/Open-Assistant) like initiative on that (@yk), but maybe you could start one with the position you're in?

There's lots of data already on user machines from your created locally held databases that... I for one, would like to use [gpt-pilot](https://github.com/Pythagora-io/gpt-pilot) MORE for if I knew that I could just say "yes, I would like to send all my saved or future conversation data to you for creating an open dataset"
